### ISSUE TITLE
Accuracy drop, and validation loss increase after commit

### ISSUE THREAD

**Issue by Str4t0n** [nnstreamer/nntrainer#877]
Running the same model and same code after pulling 31e7e9ca3181d090cf5d7e60509467549eb5224c gaves different training results:
<details>
  <summary>Before pulling 31e7e9ca3181d090cf5d7e60509467549eb5224c</summary>

```log
#1/10 - Training Loss: 5.02359 >> [ Accuracy: 8.72% - Validation Loss : 4.39116 ] 
#2/10 - Training Loss: 4.19095 >> [ Accuracy: 11.04% - Validation Loss : 4.17739 ] 
#3/10 - Training Loss: 3.99177 >> [ Accuracy: 11.44% - Validation Loss : 3.97594 ] 
#4/10 - Training Loss: 3.89871 >> [ Accuracy: 15.84% - Validation Loss : 3.78107 ] 
#5/10 - Training Loss: 3.68158 >> [ Accuracy: 15.28% - Validation Loss : 3.68737 ] 
#6/10 - Training Loss: 3.65966 >> [ Accuracy: 16.8% - Validation Loss : 3.66477 ] 
#7/10 - Training Loss: 3.55599 >> [ Accuracy: 19.84% - Validation Loss : 3.50755 ] 
#8/10 - Training Loss: 3.59489 >> [ Accuracy: 15.28% - Validation Loss : 3.82755 ] 
#9/10 - Training Loss: 3.53586 >> [ Accuracy: 18.92% - Validation Loss : 3.40347 ] 
#10/10 - Training Loss: 3.37745 >> [ Accuracy: 18.24% - Validation Loss : 3.4634 ]
```
</details>
<details>
  <summary>After pulling 31e7e9ca3181d090cf5d7e60509467549eb5224c</summary>

```log
#1/10 - Training Loss: 5.36249 >> [ Accuracy: 1.72% - Validation Loss : 25.4045 ] 
#2/10 - Training Loss: 4.11813 >> [ Accuracy: 2.28% - Validation Loss : 10.7656 ] 
#3/10 - Training Loss: 3.99825 >> [ Accuracy: 5.56% - Validation Loss : 7.75355 ] 
#4/10 - Training Loss: 3.87466 >> [ Accuracy: 11.6% - Validation Loss : 4.71803 ] 
#5/10 - Training Loss: 3.75498 >> [ Accuracy: 13.2% - Validation Loss : 4.29773 ] 
#6/10 - Training Loss: 3.70072 >> [ Accuracy: 15.24% - Validation Loss : 3.91692 ] 
#7/10 - Training Loss: 3.63354 >> [ Accuracy: 14.32% - Validation Loss : 3.9241 ] 
#8/10 - Training Loss: 3.56396 >> [ Accuracy: 14.68% - Validation Loss : 3.63199 ] 
#9/10 - Training Loss: 3.48811 >> [ Accuracy: 17.32% - Validation Loss : 3.54894 ] 
#10/10 - Training Loss: 3.44803 >> [ Accuracy: 15.16% - Validation Loss : 3.92852 ]
```
</details>
The test was run multiple times, and every time we got the same results (±1% accuracy due to the random input we are working with).

```cpp
// we tried to disable optimization like this, but the result was the same
nntrainer::NeuralNetwork baseModel(nntrainer::AppContext::Global(), false);

// and also tried to set
baseModel.setGradientMemoryOptimization(false);
baseModel.setDerivativeMemoryOptimization(false); 
baseModel.setInPlaceLayerOptimization(false);
baseModel.setInferenceInOutMemoryOptimization(false);
```

The issue is somewhat related to https://github.com/nnstreamer/nntrainer/issues/871 , because we are using the same configuration file

**Comment by zhoonit** [comment]
Thank you for posting the issue.

Is your setup includes batchnormalization layer?

There was a problem with batchnormalization that it was updating average mean and average variance on the inference mode.
After epoch there is a validation run, which would've affected the mean and variance which would result in performance change.

If your setup does not include batchnormalization, i thinks there should be a bug we have to look into. If that's the case, could you share more information about the issue?

**Comment by kparichay** [comment]
As @zhoonit , 31e7e9ca3181d090cf5d7e60509467549eb5224c changes the beahvior of the inference for batch normalization.
So, a different validation result is expected.
If you can sync the input for the results, the training loss must match. Please try your model training with the same input (it can be random but if you use the same seed, it will be deterministic) and check the training loss.

The difference in validation accuracy for the example you shared can be caused by multiple reasons:
1. Size of the training data - if the training dataset is small, the batch normalization parameters might need more than just a couple of epochs to converge, as is happening here.
2. Initialization/input variation - the variations with different initializations/inputs will have some variations ±1%.

**Comment by Str4t0n** [comment]
@kparichay , @zhoonit , thank you for your replies.
> @zhoonit: Is your setup includes batchnormalization layer?

Yes, our model contains multiple batchnormalization layer for the model training.

> @kparichay: If you can sync the input for the results, the training loss must match. 

After syncing up the seed, we got the following results :

<details>
  <summary>Before 31e7e9ca3181d090cf5d7e60509467549eb5224c :</summary>

```log
#1/10 - Training Loss: 5.20508 >> [ Accuracy: 9.68% - Validation Loss : 4.05317 ] 
#2/10 - Training Loss: 4.17794 >> [ Accuracy: 14.64% - Validation Loss : 3.70753 ] 
#3/10 - Training Loss: 3.97247 >> [ Accuracy: 18.08% - Validation Loss : 3.36693 ] 
#4/10 - Training Loss: 3.85025 >> [ Accuracy: 16.96% - Validation Loss : 3.56311 ] 
#5/10 - Training Loss: 3.73552 >> [ Accuracy: 18.92% - Validation Loss : 3.44152 ] 
#6/10 - Training Loss: 3.63949 >> [ Accuracy: 20.68% - Validation Loss : 3.38732 ] 
#7/10 - Training Loss: 3.58332 >> [ Accuracy: 24.28% - Validation Loss : 3.10129 ] 
#8/10 - Training Loss: 3.56946 >> [ Accuracy: 23.84% - Validation Loss : 3.11461 ] 
#9/10 - Training Loss: 3.53093 >> [ Accuracy: 25.28% - Validation Loss : 3.05785 ] 
#10/10 - Training Loss: 3.40488 >> [ Accuracy: 27% - Validation Loss : 3.05904 ]
```
</details>

<details>
  <summary>After 31e7e9ca3181d090cf5d7e60509467549eb5224c (main HEAD[fb0611f105b093657941e1b984902d79e3ef8f65]):</summary>

```log
#1/10 - Training Loss: 5.20508 >> [ Accuracy: 1.88% - Validation Loss : 22.965 ] 
#2/10 - Training Loss: 4.17794 >> [ Accuracy: 4.56% - Validation Loss : 9.41912 ] 
#3/10 - Training Loss: 3.97247 >> [ Accuracy: 5.52% - Validation Loss : 6.24138 ] 
#4/10 - Training Loss: 3.85025 >> [ Accuracy: 9.36% - Validation Loss : 4.93885 ] 
#5/10 - Training Loss: 3.73552 >> [ Accuracy: 12.68% - Validation Loss : 4.2429 ] 
#6/10 - Training Loss: 3.63949 >> [ Accuracy: 16.68% - Validation Loss : 3.73396 ] 
#7/10 - Training Loss: 3.58332 >> [ Accuracy: 22.68% - Validation Loss : 3.23321 ] 
#8/10 - Training Loss: 3.56946 >> [ Accuracy: 20.76% - Validation Loss : 3.35632 ] 
#9/10 - Training Loss: 3.53093 >> [ Accuracy: 22.24% - Validation Loss : 3.33128 ] 
#10/10 - Training Loss: 3.40488 >> [ Accuracy: 24.24% - Validation Loss : 3.189 ]
```
</details>

> @zhoonit: After epoch there is a validation run, which would've affected the mean and variance which would result in performance change.

So, the results that we are getting after the commit are the good ones, i assume?

**Comment by kparichay** [comment]
@Str4t0n The results look good. The training loss matches as it should and the difference in validation loss is as per @zhoonit explained.

> So, the results that we are getting after the commit are the good ones, i assume?

Yes.
